fxrecorder/src/lib/analysis.rs

```diff
@@ -0,0 +1,406 @@
+/// `directory` in the form of `directory/frames/NNNNNN.png`, where N is a six
+/// digit timestamp of each frame.
+///
+/// Not all frames are extracted. We use video filters to only extract
+/// sequentially different frames.
+pub fn extract_frames(
+    log: slog::Logger,
+    video_path: &Path,
+    target_directory: &Path,
+) -> Result<PathBuf, ExtractFramesError> {
+    let frames_dir = target_directory.join("frames");
+
+    create_dir_all(&frames_dir)
+        .map_err(|source| ExtractFramesError::CreateDir(source, frames_dir.clone()))?;
+
+    let output_format = frames_dir.join("%06d.png");
+
+    let args = vec![
+        OsStr::new("-i"),
+        video_path.as_os_str(),
+        // Pass through frames with their timestmaps from the demuxer to the muxer.
+        OsStr::new("-vsync"),
+        OsStr::new("passthrough"),
+        // Use the image2 format.
+        OsStr::new("-f"),
+        OsStr::new("image2"),
+        // Use the "presentation time stamp" (frame number from 0) in the output. The
+        // timestamp each frame can be found by multiplying the frame number by
+        // the "time base" (1 / framerate).
+        OsStr::new("-frame_pts"),
+        OsStr::new("true"),
+        // mpdecimate drops sequentially similar frames from the output. This
+        // reduces the number of rendered frames from a few thousand to around a
+        // hundred.
+        OsStr::new("-vf"),
+        OsStr::new("mpdecimate"),
+        // The output file path format.
+        output_format.as_os_str(),
+    ];
+
+    info!(log, "extracting frames"; "args" => ?&args);
+
+    run_ffmpeg(log.clone(), &args).map_err(ExtractFramesError::Ffmpeg)?;
+    Ok(frames_dir)
+}
+
+/// Information about a frame being processed in
+/// [`find_first_orange_frame`][function.find_first_orange_frame.html].
+#[derive(Debug)]
+struct FrameInfo {
+    /// The path to the frame.
+    path: PathBuf,
+
+    /// The frame number.
+    frame_num: u32,
+}
+
+/// Squared Euclidean Distance between two colours as 3-vectors.
+fn squared_distance(a: &Rgb<u8>, b: &Rgb<u8>) -> i64 {
+    let dr = a[0] as i64 - b[0] as i64;
+    let dg = a[1] as i64 - b[1] as i64;
+    let db = a[2] as i64 - b[2] as i64;
+
+    dr * dr + dg * dg + db * db
+}
+
+#[derive(Debug, Error)]
+pub enum OrangeError {
+    #[error("could not read frame directory: {}", .0)]
+    ReadDir(#[source] io::Error),
+
+    #[error("could not read file `{}': {}", .1.display(), .0)]
+    Open(#[source] io::Error, PathBuf),
+
+    #[error("could not load image `{}': {}'", .1.display(), .0)]
+    Load(#[source] ImageError, PathBuf),
+
+    #[error("no orange frame detected")]
+    MissingOrange,
+}
+
+/// Return the frame number of the first orange frame of the video.
+fn find_first_orange_frame(log: slog::Logger, frames_dir: &Path) -> Result<u32, OrangeError> {
+    // This is the orange that Splash generates and that visuametrics.py expects.
+    let orange = image::Rgb([222, 100, 13]);
```

Comment:
Nit: given that we use these values in two different locations, it may be nice to pull it up into a common constant.